### ISSUE TITLE
Frames not being written to output video

### ISSUE THREAD

**Issue by jeffski** [OpenShot/libopenshot#122]
I started noticing frames missing from rendered videos, not sure if this has always been the case or it started recently, or if it just my setup.

If you compile and run the following code (replace the path/file with your own input video):

```c++
#include "OpenShot.h"

int main(int argc, char **argv) {
	string inputVideo = "input.mp4";
	string outputVideo = "output.mp4";

	FFmpegReader inputReader(inputVideo);
	inputReader.Open();

	FFmpegWriter outputWriter(outputVideo);
	outputWriter.SetAudioOptions(true, "libmp3lame", inputReader.info.sample_rate, inputReader.info.channels, inputReader.info.channel_layout, 128000);
	outputWriter.SetVideoOptions(true, "libx264", inputReader.info.fps, 1024, 576, Fraction(1,1), false, false, 3000000);
	outputWriter.Open();

	for (long int frame = 1; frame <= 25; frame++)
	{
		int frame_number = frame;
		std::shared_ptr<Frame> f = inputReader.GetFrame(frame_number);
		outputWriter.WriteFrame(f);
	}

	outputWriter.Close();
	inputReader.Close();

	FFmpegReader outputReader(outputVideo);
	outputReader.Open();
	outputReader.DisplayInfo();
	outputReader.Close();

	return 0;
}
```
I would expect to have a video with 25 frames and be 1 second long. Instead I have just 5 frames.

The output from `outputReader.DisplayInfo();` is below showing 5 frames:

```
[libx264 @ 0x562b0841ce80] using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX
[libx264 @ 0x562b0841ce80] profile High, level 3.1
[libx264 @ 0x562b0841ce80] 264 - core 152 r2854 e9a5903 - H.264/MPEG-4 AVC codec - Copyleft 2003-2017 - http://www.videolan.org/x264.html - options: cabac=1 ref=3 deblock=1:0:0 analyse=0x3:0x113 me=hex subme=7 psy=1 psy_rd=1.00:0.00 mixed_ref=1 me_range=16 chroma_me=1 trellis=1 8x8dct=1 cqm=0 deadzone=21,11 fast_pskip=1 chroma_qp_offset=-2 threads=8 lookahead_threads=1 sliced_threads=0 nr=0 decimate=1 interlaced=0 bluray_compat=0 constrained_intra=0 bframes=0 weightp=2 keyint=12 keyint_min=1 scenecut=40 intra_refresh=0 rc_lookahead=12 rc=abr mbtree=1 bitrate=3000 ratetol=1.0 qcomp=0.60 qpmin=2 qpmax=30 qpstep=4 ip_ratio=1.40 aq=1:1.00
[libx264 @ 0x562b0841ce80] frame I:1     Avg QP:21.31  size:138282
[libx264 @ 0x562b0841ce80] frame P:4     Avg QP:23.27  size:  8717
[libx264 @ 0x562b0841ce80] mb I  I16..4:  0.0% 53.0% 47.0%
[libx264 @ 0x562b0841ce80] mb P  I16..4:  0.0%  0.0%  0.0%  P16..4: 32.2% 20.9% 19.6%  0.0%  0.0%    skip:27.3%
[libx264 @ 0x562b0841ce80] final ratefactor: 27.66
[libx264 @ 0x562b0841ce80] 8x8 transform intra:53.0% inter:61.8%
[libx264 @ 0x562b0841ce80] coded y,uvDC,uvAC intra: 98.6% 99.9% 98.0% inter: 21.9% 31.9% 4.1%
[libx264 @ 0x562b0841ce80] i16 v,h,dc,p:  0%  0%  0% 100%
[libx264 @ 0x562b0841ce80] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 12% 13% 10%  7% 12% 11% 11% 10% 14%
[libx264 @ 0x562b0841ce80] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 12% 13% 10%  9% 13% 11% 11% 10% 12%
[libx264 @ 0x562b0841ce80] i8c dc,h,v,p: 53% 20% 15% 13%
[libx264 @ 0x562b0841ce80] Weighted P-Frames: Y:0.0% UV:0.0%
[libx264 @ 0x562b0841ce80] ref P L0: 62.8% 11.2% 19.9%  6.1%
[libx264 @ 0x562b0841ce80] kb/s:8302.85
[libmp3lame @ 0x562b0841de40] 3 frames left in the queue on closing
----------------------------
----- File Information -----
----------------------------
--> Has Video: true
--> Has Audio: true
--> Has Single Image: false
--> Duration: 0.86 Seconds
--> File Size: 0.17 MB
----------------------------
----- Video Attributes -----
----------------------------
--> Width: 1024
--> Height: 576
--> Pixel Format: 0
--> Frames Per Second: 39.96 (40000/1001)
--> Video Bit Rate: 1839 kb/s
--> Pixel Ratio: 1.00 (1/1)
--> Display Aspect Ratio: 1.78 (16/9)
--> Video Codec: h264
--> Video Length: 5 Frames
--> Video Stream Index: 0
--> Video Timebase: 0.00 (1/30000)
--> Interlaced: false
--> Interlaced: Top Field First: true
----------------------------
----- Audio Attributes -----
----------------------------
--> Audio Codec: mp3
--> Audio Bit Rate: 128 kb/s
--> Sample Rate: 48000 Hz
--> # of Channels: 2
--> Channel Layout: 3
--> Audio Stream Index: 1
--> Audio Timebase: 0.00 (1/48000)
----------------------------
--------- Metadata ---------
----------------------------
--> compatible_brands: isomiso2avc1mp41
--> encoder: Lavf57.83.100
--> handler_name: SoundHandler
--> language: und
--> major_brand: isom
--> minor_version: 512
```

If anyone can test and confirm that would be great.

**Comment by jeffski** [comment]
Possibly related to: https://github.com/OpenShot/libopenshot/issues/115

**Comment by jeffski** [comment]
One thing that is consistent with this bug is that I am losing 20 frames every render.

I had a dig round the code and noticed the magic number 20 seems to be being used for caching and/or multi-threading. I haven't had the time to figure out what this is all doing but perhaps the problem lies around that.

Here are some places around caching, threading and the mysterious 20:

https://github.com/OpenShot/libopenshot/blob/bbbf42aeabf6aeea6867d87883fc8c9c807adbaa/src/AudioResampler.cpp#L33

https://github.com/OpenShot/libopenshot/blob/bbbf42aeabf6aeea6867d87883fc8c9c807adbaa/tests/Cache_Tests.cpp#L79

https://github.com/OpenShot/libopenshot/blob/bbbf42aeabf6aeea6867d87883fc8c9c807adbaa/src/CacheMemory.cpp#L308

https://github.com/OpenShot/libopenshot/blob/bbbf42aeabf6aeea6867d87883fc8c9c807adbaa/src/DecklinkOutput.cpp#L261

https://github.com/OpenShot/libopenshot/blob/bbbf42aeabf6aeea6867d87883fc8c9c807adbaa/src/CacheDisk.cpp#L453

https://github.com/OpenShot/libopenshot/blob/694ee7c24ec21bd71583f5e3c554463cd9233884/src/FFmpegReader.cpp#L477

https://github.com/OpenShot/libopenshot/blob/694ee7c24ec21bd71583f5e3c554463cd9233884/src/FFmpegReader.cpp#L826

https://github.com/OpenShot/libopenshot/blob/bbbf42aeabf6aeea6867d87883fc8c9c807adbaa/src/FrameMapper.cpp#L484

**Comment by jeffski** [comment]
Did some debugging tonight and the first 20 frames error here:

https://github.com/OpenShot/libopenshot/blob/628eb373079e6eb6c6eb4ecdc2b9abf9d4e26f58/src/FFmpegWriter.cpp#L1624-L1628

`ret` is coming back with an error code -11 (Try Again) so the frame doesn't seem to be written here:
https://github.com/OpenShot/libopenshot/blob/628eb373079e6eb6c6eb4ecdc2b9abf9d4e26f58/src/FFmpegWriter.cpp#L1683 

The final 5 frames work fine.

Will keep investigating. If this looks familiar or anyone understands what is going on here I'd welcome some pointers.

**Comment by peanutbutterandcrackers** [comment]
ping @ferdnyc, @N3WWN, @DylanC, @jonoomph, @steinybot

**Comment by jeffski** [comment]
I did some more investigation in to this and tried a bunch of things. The main thing I discovered was changing the codec of the FFmpegWriter affects the number of frames written.

**x264**
Video Length: 5 Frames

**x265**
Video Length: 20 Frames
But all frames are blank. I tried 100 frames it say 67 frames and includes about 2 seconds of actual visible frames, the rest are blank.

**xvid**
Video Length: 24 Frames
It actually works correctly!! but is slower and quality is not as good.

I also tried to turn off seeking in FFmpegReader as the code mentions it is not reliable across codecs and may add blank images:

https://github.com/OpenShot/libopenshot/blob/628eb373079e6eb6c6eb4ecdc2b9abf9d4e26f58/include/FFmpegReader.h#L205-L206

https://github.com/OpenShot/libopenshot/blob/628eb373079e6eb6c6eb4ecdc2b9abf9d4e26f58/include/FFmpegReader.h#L221-L223

But this made no difference.

I also tried ffmpeg 2.2 and 3.4 but no difference.

This problem seems to complicated for me to work on right now so I am going to leave things here. I'm certain there is an issue and it would be great if it could be fixed.

For now I am just going with a workaround and will add 20 frames to my output videos which should allow the entire video to render.

**Comment by jeffski** [comment]
Actually, I just checked some old videos and they all work fine. I think this issue was introduced in the last 6 months. There are a few commits in the last 6 months and around seeking, caching and blank frames:
https://github.com/OpenShot/libopenshot/commits/develop/src/FFmpegReader.cpp

I'll try and identify the commit where things broke for me another night.

**Comment by DylanC** [comment]
@jonoomph - Do we know where in the code this broke? Something for Peter to look at perhaps?

**Comment by jeffski** [comment]
I think the problem is to do with how encoding works, is the end of stream being flushed/drained correctly? From the docs: https://www.ffmpeg.org/doxygen/3.4/group__lavc__encdec.html

It says: 
End of stream situations. These require "flushing" (aka draining) the codec, as the codec might buffer multiple frames or packets internally for performance or out of necessity (consider B-frames). This is handled as follows:

- Instead of valid input, send NULL to the avcodec_send_packet() (decoding) or avcodec_send_frame() (encoding) functions. This will enter draining mode.
- Call avcodec_receive_frame() (decoding) or avcodec_receive_packet() (encoding) in a loop until AVERROR_EOF is returned. The functions will not return AVERROR(EAGAIN), unless you forgot to enter draining mode.
- Before decoding can be resumed again, the codec has to be reset with avcodec_flush_buffers().

Not sure if the library is follow this process correctly. I think this is being handled here but not sure if it correct: 

https://github.com/OpenShot/libopenshot/blob/628eb373079e6eb6c6eb4ecdc2b9abf9d4e26f58/src/FFmpegWriter.cpp#L1624-L1628

**Comment by jeffski** [comment]
Some more digging... It looks like the flushing is started in `void FFmpegWriter::flush_encoders()` here:

https://github.com/OpenShot/libopenshot/blob/22384c770501b9d737501847400f571f67edf385/src/FFmpegWriter.cpp#L579-L585

But there is no corresponding `avcodec_receive_packet()` with the flushed frames and no `av_interleaved_write_frame()` to write them.

I think another loop somewhere to receive and write the encoded frames is needed.

**Comment by jeffski** [comment]
I had a quick look at trying to fix this today but I think I am better leaving in this in more experienced, capable hands.

I think I've confirmed there is a bug here, the cause and where to start looking so hopefully that will give someone a head start.

**Comment by DylanC** [comment]
@jeffski - Peter is the guy who originally wrote the code and has submitted a patch to Jonathan. Thanks for looking into it anyway.

**Comment by jeffski** [comment]
Thanks @DylanC, good to hear, be great to have this fixed.

**Comment by peanutbutterandcrackers** [comment]
Mr. Peter has sent me a patch (well, I requested it). Perhaps you might want to test it out. He has sent it originally to Mr. Thomas, who will probably apply it upstream as soon as he gets back in action.
[lastframecorrect.patch.tar.gz](https://github.com/OpenShot/libopenshot/files/2187387/lastframecorrect.patch.tar.gz)

**Comment by jeffski** [comment]
Yes!! It appears to have done the trick, works as expected.

**Comment by peanutbutterandcrackers** [comment]
Awesome. It was, after all, written by Mr. Peter himself. :)

Now we'll just have to wait for @jonoomph to apply the patch upstream. :)

**Comment by N3WWN** [comment]
Is this patch available in a build somewhere or do I need to apply it to my checked out code manually?  The reason I ask is that I am wondering if it addresses https://github.com/OpenShot/openshot-qt/issues/733 where there is one missing frame from the start of rendered videos.

**Comment by DylanC** [comment]
@N3WWN - Needs to be applied manually. Someone could however put it in a PR and we could merge perhaps.

**Comment by jeffski** [comment]
I have taken the liberty of creating a pull request for this: https://github.com/OpenShot/libopenshot/pull/145 - thanks to Peter for the patch.

**Comment by jeffski** [comment]
This has now been resolved in PR #145
